crates/services/src/services/share.rs

```diff
@@ -386,61 +390,94 @@ pub(super) async fn sync_local_task_for_shared_task(
-                shared_task.description.clone(),
-                shared_task.status.clone(),
-                task.parent_task_attempt,
-            )
-            .await?;
-        }
+    let project_id = match shared_task.project_id {
+        Some(project_id) => project_id,
+        None => return Ok(()),
+    };
 
-        return Ok(());
-    }
+    let create_task_if_not_exists = {
+        let assignee_is_current_user = matches!(
+            (shared_task.assignee_user_id.as_deref(), current_user_id),
+            (Some(assignee), Some(current)) if assignee == current
+        );
+        let creator_is_current_user = matches!((creator_user_id, current_user_id), (Some(creator), Some(current)) if creator == current);
 
-    let assignee_is_current_user = match (shared_task.assignee_user_id.as_deref(), current_user_id)
-    {
-        (Some(assignee), Some(current)) => assignee == current,
-        _ => false,
+        assignee_is_current_user && !creator_is_current_user
     };
 
-    if !assignee_is_current_user {
+    Task::sync_from_shared_task(
+        pool,
+        shared_task.id,
+        project_id,
+        shared_task.title.clone(),
+        shared_task.description.clone(),
+        shared_task.status.clone(),
+        create_task_if_not_exists,
+    )
+    .await?;
+
+    Ok(())
+}
+
+pub async fn link_shared_tasks_to_project(
+    pool: &SqlitePool,
+    sessions: &ClerkSessionStore,
+    project_id: Uuid,
+    github_repo_id: i64,
+) -> Result<(), ShareError> {
+    if GithubProjectLinkingCache::already_linked(github_repo_id) {
         return Ok(());
     }
 
-    let creator_is_current_user = match (creator_user_id, current_user_id) {
-        (Some(creator), Some(current)) => creator == current,
-        _ => false,
-    };
+    let linked_tasks =
+        SharedTask::link_to_project_by_repo_id(pool, github_repo_id, project_id).await?;
 
-    if creator_is_current_user {
-        // Current user created the shared task but has no corresponding local shared-task record.
-        // This can happen if a share acivity event is received before the task sharing operation completes.
+    if linked_tasks.is_empty() {
+        GithubProjectLinkingCache::cache_as_linked(github_repo_id);
         return Ok(());
     }
 
-    // Current user owns the shared task but has no local record; create and link one.
-    let create = CreateTask::from_shared_task(
-        shared_task.project_id,
-        shared_task.title.clone(),
-        shared_task.description.clone(),
-        shared_task.status.clone(),
-        shared_task.id,
-    );
-    let task_id = Uuid::new_v4();
-    Task::create(pool, &create, task_id).await?;
+    let current_user_id = sessions.last().await.as_ref().map(|s| s.user_id.clone());
+
+    for task in linked_tasks {
+        sync_local_task_for_shared_task(pool, &task, current_user_id.as_deref(), None).await?;
+    }
+
+    GithubProjectLinkingCache::cache_as_linked(github_repo_id);
 
     Ok(())
 }
+
```

Comment:
Do we really need this cache? We can avoid redundant linking operations at a higher level (e.g. in `refresh_remote_metadata`) and even if there is redundancy, it's not that expensive.